File: C10_Lights/LightSequencerCS/Sequence.cs
//////////////////////////////////////////////////////////////////////////////////
//	Sequence.cs
//	Light Sequencer
//	Written by Brian Peek (http://www.brianpeek.com/)
//	for the Animated Holiday Lights article
//		at Coding4Fun (http://msdn.microsoft.com/coding4fun/)
//////////////////////////////////////////////////////////////////////////////////

using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace LightSequencer
{
	public enum MusicType
	{
		Sample = 0,
		MIDI = 1
	}

	public class Sequence
	{
		private const string FILE_VERSION = "2";

		public string Title;
		public string Artist;
		public string MusicFile;
		public int MusicLength;
		public float Interval;	// in ms
		public int Version = int.Parse(FILE_VERSION);
		public List<Channel> Channels = new List<Channel>();
		public MusicType MusicType;

		public Sequence() : this(null)
		{
		}

		public Sequence(string filename)
		{
			if(!string.IsNullOrEmpty(filename))
				Load(filename);
		}

		public void Save(string filename)
		{
			// indent our output file
			XmlWriterSettings settings = new XmlWriterSettings();
			settings.Indent = true;

			// create a new XML file
			XmlWriter xml = XmlWriter.Create(filename, settings);

			// write start tags
			xml.WriteStartDocument();

			// save the appropriate version
			switch(this.Version)
			{
				case 1:
					SaveV1Sequence(xml);	
					break;
				default:
					SaveSequence(xml);
					break;
			}

			// write end tags
			xml.WriteEndDocument();
		
			// close it up
			xml.Flush();
			xml.Close();
		}

		private void SaveV1Sequence(XmlWriter xml)
		{
			StringBuilder sb = new StringBuilder();

			xml.WriteComment("Sequence file for LightSequencer");

			xml.WriteStartElement("sequence");

				// write base config info
				xml.WriteStartElement("config");
					xml.WriteElementString("fileVersion", "1");
					xml.WriteElementString("numChannels", this.Channels.Count.ToString());
					xml.WriteElementString("musicFile", this.MusicFile);
					xml.WriteElementString("musicLength", this.MusicLength.ToString());
				xml.WriteEndElement();

				xml.WriteStartElement("channels");

				foreach(Channel channel in this.Channels)
				{
					// 1==on, 0==off...lame, but it works. :)
					foreach(bool b in channel.Data)
						sb.Append((b ? "1" : "0"));

					// write channel specific info
					xml.WriteStartElement("channel");
						xml.WriteAttributeString("number", channel.Number.ToString());
						xml.WriteAttributeString("serialNumber", channel.SerialNumber.ToString());
						xml.WriteAttributeString("outputIndex", channel.OutputIndex.ToString());
						xml.WriteString(sb.ToString());
					xml.WriteEndElement();

					sb.Remove(0, sb.Length);
				}

				xml.WriteEndElement();

			xml.WriteEndElement();
		}

		private void SaveSequence(XmlWriter xml)
		{
			StringBuilder sb = new StringBuilder();

			xml.WriteComment("Sequence file for LightSequencer");

			xml.WriteStartElement("sequence");

				// write base config info
				xml.WriteStartElement("config");
					xml.WriteElementString("fileVersion", FILE_VERSION);
					xml.WriteElementString("musicFile", this.MusicFile);
					xml.WriteElementString("musicLength", this.MusicLength.ToString());
					xml.WriteElementString("numChannels", this.Channels.Count.ToString());
					xml.WriteElementString("interval", this.Interval.ToString());
				xml.WriteEndElement();

				// write music information
				xml.WriteStartElement("musicInfo");
					xml.WriteElementString("title", this.Title);
					xml.WriteElementString("artist", this.Artist);
				xml.WriteEndElement();

				// write channels
				xml.WriteStartElement("channels");

				foreach(Channel channel in this.Channels)
				{
					// 1==on, 0==off...lame, but it works. :)
					foreach(bool b in channel.Data)
						sb.Append((b ? "1" : "0"));

					// write channel specific info
					xml.WriteStartElement("channel");
						xml.WriteAttributeString("number", channel.Number.ToString());
						xml.WriteAttributeString("serialNumber", channel.SerialNumber.ToString());
						xml.WriteAttributeString("outputIndex", channel.OutputIndex.ToString());
						xml.WriteAttributeString("midiChannel", channel.MIDIChannel.ToString());
						xml.WriteString(sb.ToString());
					xml.WriteEndElement();

					sb.Remove(0, sb.Length);
				}
				xml.WriteEndElement();
			xml.WriteEndElement();
		}

		public void Load(string filename)
		{
			// load the XML file
			XmlDocument xmlDoc = new XmlDocument();
			xmlDoc.Load(filename);

			// grab the base config info
			this.Version = int.Parse(xmlDoc.SelectSingleNode("sequence/config/fileVersion").InnerText);
			switch(this.Version)
			{
				case 1:
					LoadV1Sequence(xmlDoc);
					break;
				default:
					LoadSequence(xmlDoc);
					break;
			}
		}

		public void LoadV1Sequence(XmlDocument xmlDoc)
		{
			XmlNode	node;

			this.Interval = 50;

			int numChannels = int.Parse(xmlDoc.SelectSingleNode("sequence/config/numChannels").InnerText);
			this.MusicFile = xmlDoc.SelectSingleNode("sequence/config/musicFile").InnerText;
			this.MusicLength = int.Parse(xmlDoc.SelectSingleNode("sequence/config/musicLength").InnerText);
			this.MusicType = MusicType.Sample;

			this.Channels.Clear();

			// add each channel to our internal list for easy tracking
			for(int i = 0; i < numChannels; i++)
			{
				node = xmlDoc.SelectSingleNode("sequence/channels/channel[@number='" + i.ToString() + "']");
				if(node != null)
				{
					int serial = int.Parse(node.Attributes["serialNumber"].Value);
					char[] data = node.InnerText.ToCharArray();
					this.Channels.Add(new Channel(i, serial, int.Parse(node.Attributes["outputIndex"].Value), data.Length));
					for(int j = 0; j < data.Length; j++)
						this.Channels[i].Data[j] = (data[j] == '1');
				}
				else
					this.Channels.Add(new Channel(i, -1, -1));
			}
		}

		public void LoadSequence(XmlDocument xmlDoc)
		{
			XmlNode	node;

			// grab the base config info
			this.MusicFile = xmlDoc.SelectSingleNode("sequence/config/musicFile").InnerText;

			if(this.MusicFile.EndsWith(".mid"))
				this.MusicType = MusicType.MIDI;

			this.MusicLength = int.Parse(xmlDoc.SelectSingleNode("sequence/config/musicLength").InnerText);
			this.Interval = float.Parse(xmlDoc.SelectSingleNode("sequence/config/interval").InnerText);

			int numChannels = int.Parse(xmlDoc.SelectSingleNode("sequence/config/numChannels").InnerText);

			this.Title = xmlDoc.SelectSingleNode("sequence/musicInfo/title").InnerText;
			this.Artist = xmlDoc.SelectSingleNode("sequence/musicInfo/artist").InnerText;

			this.Channels.Clear();

			// add each channel to our internal list for easy tracking
			for(int i = 0; i < numChannels; i++)
			{
				node = xmlDoc.SelectSingleNode("sequence/channels/channel[@number='" + i.ToString() + "']");
				if(node != null)
				{
					int serial = int.Parse(node.Attributes["serialNumber"].Value);
					char[] data = node.InnerText.ToCharArray();

					// add the channel to the channel list
					this.Channels.Add(new Channel(i, serial, int.Parse(node.Attributes["outputIndex"].Value), int.Parse(node.Attributes["midiChannel"].Value), data.Length));

					// read out the data
					for(int j = 0; j < data.Length; j++)
						this.Channels[i].Data[j] = (data[j] == '1');
				}
				else
					this.Channels.Add(new Channel(i));
			}
		}
	}
}

File: C10_Lights/LightSequencerVB/Sequence.vb
'////////////////////////////////////////////////////////////////////////////////
'	Sequence.cs
'	Light Sequencer
'	Written by Brian Peek (http://www.brianpeek.com/)
'	for the Animated Holiday Lights article
'		at Coding4Fun (http://msdn.microsoft.com/coding4fun/)
'////////////////////////////////////////////////////////////////////////////////


Imports Microsoft.VisualBasic
Imports System.Collections.Generic
Imports System.Text
Imports System.Xml

Namespace LightSequencer
	Public Enum MusicType
		Sample = 0
		MIDI = 1
	End Enum

	Public Class Sequence
		Private Const FILE_VERSION As String = "2"

		Public Title As String
		Public Artist As String
		Public MusicFile As String
		Public MusicLength As Integer
		Public Interval As Single ' in ms
		Public Version As Integer = Integer.Parse(FILE_VERSION)
		Public Channels As List(Of Channel) = New List(Of Channel)()
		Public MusicType As MusicType

		Public Sub New()
			Me.New(Nothing)
		End Sub

		Public Sub New(ByVal filename As String)
			If (Not String.IsNullOrEmpty(filename)) Then
				Load(filename)
			End If
		End Sub

		Public Sub Save(ByVal filename As String)
			' indent our output file
			Dim settings As New XmlWriterSettings()
			settings.Indent = True

			' create a new XML file
			Dim xml As XmlWriter = XmlWriter.Create(filename, settings)

			' write start tags
			xml.WriteStartDocument()

			' save the appropriate version
			Select Case Me.Version
				Case 1
					SaveV1Sequence(xml)
				Case Else
					SaveSequence(xml)
			End Select

			' write end tags
			xml.WriteEndDocument()

			' close it up
			xml.Flush()
			xml.Close()
		End Sub

		Private Sub SaveV1Sequence(ByVal xml As XmlWriter)
			Dim sb As New StringBuilder()

			xml.WriteComment("Sequence file for LightSequencer")

			xml.WriteStartElement("sequence")

				' write base config info
				xml.WriteStartElement("config")
					xml.WriteElementString("fileVersion", "1")
					xml.WriteElementString("numChannels", Me.Channels.Count.ToString())
					xml.WriteElementString("musicFile", Me.MusicFile)
					xml.WriteElementString("musicLength", Me.MusicLength.ToString())
				xml.WriteEndElement()

				xml.WriteStartElement("channels")

				For Each channel As Channel In Me.Channels
					' 1==on, 0==off...lame, but it works. :)
					For Each b As Boolean In channel.Data
						If b Then
							sb.Append(("1"))
						Else
							sb.Append(("0"))
						End If
					Next b

					' write channel specific info
					xml.WriteStartElement("channel")
						xml.WriteAttributeString("number", channel.Number.ToString())
						xml.WriteAttributeString("serialNumber", channel.SerialNumber.ToString())
						xml.WriteAttributeString("outputIndex", channel.OutputIndex.ToString())
						xml.WriteString(sb.ToString())
					xml.WriteEndElement()

					sb.Remove(0, sb.Length)
				Next channel

				xml.WriteEndElement()

			xml.WriteEndElement()
		End Sub

		Private Sub SaveSequence(ByVal xml As XmlWriter)
			Dim sb As New StringBuilder()

			xml.WriteComment("Sequence file for LightSequencer")

			xml.WriteStartElement("sequence")

				' write base config info
				xml.WriteStartElement("config")
					xml.WriteElementString("fileVersion", FILE_VERSION)
					xml.WriteElementString("musicFile", Me.MusicFile)
					xml.WriteElementString("musicLength", Me.MusicLength.ToString())
					xml.WriteElementString("numChannels", Me.Channels.Count.ToString())
					xml.WriteElementString("interval", Me.Interval.ToString())
				xml.WriteEndElement()

				' write music information
				xml.WriteStartElement("musicInfo")
					xml.WriteElementString("title", Me.Title)
					xml.WriteElementString("artist", Me.Artist)
				xml.WriteEndElement()

				' write channels
				xml.WriteStartElement("channels")

				For Each channel As Channel In Me.Channels
					' 1==on, 0==off...lame, but it works. :)
					For Each b As Boolean In channel.Data
						If b Then
							sb.Append(("1"))
						Else
							sb.Append(("0"))
						End If
					Next b

					' write channel specific info
					xml.WriteStartElement("channel")
						xml.WriteAttributeString("number", channel.Number.ToString())
						xml.WriteAttributeString("serialNumber", channel.SerialNumber.ToString())
						xml.WriteAttributeString("outputIndex", channel.OutputIndex.ToString())
						xml.WriteAttributeString("midiChannel", channel.MIDIChannel.ToString())
						xml.WriteString(sb.ToString())
					xml.WriteEndElement()

					sb.Remove(0, sb.Length)
				Next channel
				xml.WriteEndElement()
			xml.WriteEndElement()
		End Sub

		Public Sub Load(ByVal filename As String)
			' load the XML file
			Dim xmlDoc As New XmlDocument()
			xmlDoc.Load(filename)

			' grab the base config info
			Me.Version = Integer.Parse(xmlDoc.SelectSingleNode("sequence/config/fileVersion").InnerText)
			Select Case Me.Version
				Case 1
					LoadV1Sequence(xmlDoc)
				Case Else
					LoadSequence(xmlDoc)
			End Select
		End Sub

		Public Sub LoadV1Sequence(ByVal xmlDoc As XmlDocument)
			Dim node As XmlNode

			Me.Interval = 50

			Dim numChannels As Integer = Integer.Parse(xmlDoc.SelectSingleNode("sequence/config/numChannels").InnerText)
			Me.MusicFile = xmlDoc.SelectSingleNode("sequence/config/musicFile").InnerText
			Me.MusicLength = Integer.Parse(xmlDoc.SelectSingleNode("sequence/config/musicLength").InnerText)
			Me.MusicType = MusicType.Sample

			Me.Channels.Clear()

			' add each channel to our internal list for easy tracking
			For i As Integer = 0 To numChannels - 1
				node = xmlDoc.SelectSingleNode("sequence/channels/channel[@number='" & i.ToString() & "']")
				If node IsNot Nothing Then
					Dim serial As Integer = Integer.Parse(node.Attributes("serialNumber").Value)
					Dim data() As Char = node.InnerText.ToCharArray()
					Me.Channels.Add(New Channel(i, serial, Integer.Parse(node.Attributes("outputIndex").Value), data.Length))
					For j As Integer = 0 To data.Length - 1
						Me.Channels(i).Data(j) = (data(j) = "1"c)
					Next j
				Else
					Me.Channels.Add(New Channel(i, -1, -1))
				End If
			Next i
		End Sub

		Public Sub LoadSequence(ByVal xmlDoc As XmlDocument)
			Dim node As XmlNode

			' grab the base config info
			Me.MusicFile = xmlDoc.SelectSingleNode("sequence/config/musicFile").InnerText

			If Me.MusicFile.EndsWith(".mid") Then
				Me.MusicType = MusicType.MIDI
			End If

			Me.MusicLength = Integer.Parse(xmlDoc.SelectSingleNode("sequence/config/musicLength").InnerText)
			Me.Interval = Single.Parse(xmlDoc.SelectSingleNode("sequence/config/interval").InnerText)

			Dim numChannels As Integer = Integer.Parse(xmlDoc.SelectSingleNode("sequence/config/numChannels").InnerText)

			Me.Title = xmlDoc.SelectSingleNode("sequence/musicInfo/title").InnerText
			Me.Artist = xmlDoc.SelectSingleNode("sequence/musicInfo/artist").InnerText

			Me.Channels.Clear()

			' add each channel to our internal list for easy tracking
			For i As Integer = 0 To numChannels - 1
				node = xmlDoc.SelectSingleNode("sequence/channels/channel[@number='" & i.ToString() & "']")
				If node IsNot Nothing Then
					Dim serial As Integer = Integer.Parse(node.Attributes("serialNumber").Value)
					Dim data() As Char = node.InnerText.ToCharArray()

					' add the channel to the channel list
					Me.Channels.Add(New Channel(i, serial, Integer.Parse(node.Attributes("outputIndex").Value), Integer.Parse(node.Attributes("midiChannel").Value), data.Length))

					' read out the data
					For j As Integer = 0 To data.Length - 1
						Me.Channels(i).Data(j) = (data(j) = "1"c)
					Next j
				Else
					Me.Channels.Add(New Channel(i))
				End If
			Next i
		End Sub
	End Class
End Namespace
